### PR TITLE
[routing] Fix unneeded loops in LeapsOnly mode

### PR DESCRIPTION
**Pull request by tatiana-kondakova**

### CODE REVIEW

**Comment by tatiana-kondakova** [comment]
JTTP

**Comment by tatiana-kondakova** [comment]
JTA

File: routing/index_router.cpp [hunk 1]
```diff
@@ -725,41 +725,62 @@ IRouter::ResultCode IndexRouter::ProcessLeaps(vector<Segment> const & input,
...
 
   WorldGraph & worldGraph = starter.GetGraph();
 
   // For all leaps except the first leap which connects start to mwm exit in LeapsOnly mode we need
   // to drop first segment of the leap because we already have its twin from the previous mwm.
   bool dropFirstSegment = false;
-  for (size_t i = 0; i < input.size(); ++i)
+
+  // While route building in LeapOnly mode we estimate start to mwm exit distance. This estimation may
+  // be incorrect and it causes apearence of unneeded loops: route goes from start to wrongly selected mwm exit,
```
Comment (mpimenov) [inline routing/index_router.cpp]: s/apearence/appearance/

File: routing/index_router.cpp [hunk 2]
```diff
@@ -725,41 +725,62 @@ IRouter::ResultCode IndexRouter::ProcessLeaps(vector<Segment> const & input,
...
+  // then we have small uneeded leap in other mwm which returns us to start mwm, then we have leap in start mwm
+  // to the correct start mwm exit and then we have normal route.
+  // |input| mwm ids for such case look like
+  // { fake, startId, otherId, otherId, startId, startId, .. pairs of ids for other leaps .. , finishId, fake}.
+  // To avoid these behavior we collapse all leaps from start to last occurrence of startId to one leap and
```
Comment (mpimenov) [inline routing/index_router.cpp]: s/these/this/

File: routing/index_router.cpp [hunk 3]
```diff
@@ -725,41 +725,62 @@ IRouter::ResultCode IndexRouter::ProcessLeaps(vector<Segment> const & input,
...
+  auto const startLeapEndReverseIt = find_if(input.rbegin() + 2, input.rend(),
+                                             [firstMwmId](Segment const & s) { return s.GetMwmId() == firstMwmId; });
+  auto const startLeapEndIt = startLeapEndReverseIt.base() - 1;
+  auto const startLeapEnd = distance(input.begin(), startLeapEndIt);
+
+  // The last leap processed the same way. See the comment above.
+  auto const lastMwmId = input[input.size() - 2].GetMwmId();
+  auto const finishLeapStartIt = find_if(startLeapEndIt, input.end(),
+                                         [lastMwmId](Segment const & s) { return s.GetMwmId() == lastMwmId; });
+  size_t finishLeapStart = distance(input.begin(), finishLeapStartIt);
```
Comment (bykoianko) [inline routing/index_router.cpp]: const?

File: routing/index_router.cpp (hunk 1 above)
Comment (tatiana-kondakova) [reply routing/index_router.cpp]: done

File: routing/index_router.cpp (hunk 2 above)
Comment (tatiana-kondakova) [reply routing/index_router.cpp]: done

File: routing/index_router.cpp (hunk 3 above)
Comment (tatiana-kondakova) [reply routing/index_router.cpp]: done

File: routing/index_router.cpp [hunk 4]
```diff
@@ -725,41 +725,62 @@ IRouter::ResultCode IndexRouter::ProcessLeaps(vector<Segment> const & input,
...
   WorldGraph & worldGraph = starter.GetGraph();
 
   // For all leaps except the first leap which connects start to mwm exit in LeapsOnly mode we need
   // to drop first segment of the leap because we already have its twin from the previous mwm.
   bool dropFirstSegment = false;
-  for (size_t i = 0; i < input.size(); ++i)
+
+  // While route building in LeapOnly mode we estimate start to mwm exit distance. This estimation may
+  // be incorrect and it causes appearance of unneeded loops: route goes from start to wrongly selected mwm exit,
+  // then we have small uneeded leap in other mwm which returns us to start mwm, then we have leap in start mwm
```
Comment (bykoianko) [inline routing/index_router.cpp]: unneeded

Comment (tatiana-kondakova) [reply routing/index_router.cpp]: done

**Pull request merged**